idaes/generic_models/unit_models/skeleton_model.py

```diff
@@ -0,0 +1,154 @@
+callback from the user.
+"""
+
+from pyomo.network import Port
+from pyomo.common.config import ConfigValue, In
+
+from idaes.core.process_base import declare_process_block_class, \
+    ProcessBlockData
+from idaes.core.util.exceptions import ConfigurationError
+from idaes.core.util.model_statistics import degrees_of_freedom
+from idaes.core.util import get_solver
+import idaes.logger as idaeslog
+
+__author__ = "Jaffer Ghouse"
+
+# Set up logger
+_log = idaeslog.getLogger(__name__)
+
+
+@declare_process_block_class("SkeletonUnitModel")
+class SkeletonUnitModelData(ProcessBlockData):
+    """
+    This is the class for a skeleteon unit model. This will provide a user
+    with a generic skeleton to add a custom or a surrogate unit model
+    when controlvolumes are not required.
+    """
+
+    # Create Class ConfigBlock
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare("dynamic", ConfigValue(
+        default=False,
+        domain=In([False]),
+        description="Dynamic model flag",
+        doc="""Model does not handle dynamics but should be set by user."""))
+    CONFIG.declare("custom_initializer", ConfigValue(
+        default=False,
+        domain=In([True, False]),
+        description="Custom initializer flag",
+        doc="""Boolean for custom initiliazer provided by the user"""))
+
+    def build(self):
+        """
+        General build method for SkeletonUnitModelBlockData.
+
+        Inheriting models should call `super().build`.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        super().build()
+
+        if self.config.custom_initializer:
+            self.custom_initializer_func = None
+
+    def add_ports(self, name, member_dict, doc=None):
+        """
+        This is a method to build Port objects in the SkeletonUnitModel and
+        populate them with appropriate port members as specified. User can add
+        as many inlet and outlet ports as required.
+
+        Keyword Args:
+            name : name to use for Port object.
+            dict : dictionary containing variables to be added to the port
+            doc : doc string for Port object
+
+        Returns:
+            A Pyomo Port object and associated components.
+        """
+        # Validate that member_dict is a dict
+        if not isinstance(member_dict, dict):
+            raise ConfigurationError(
+                "member_dict should be a dictionary "
+                "with the keys being the name assigned(strings) and "
+                "values being the variable objects declared in the model.")
+
+        # Create empty Port
+        p = Port(noruleinit=True, doc=doc)
+
+        # Add port object to model
+        setattr(self, name, p)
+
+        # Populate port and map names to actual variables as defined
+        for k in member_dict.keys():
+            p.add(member_dict[k], name=k)
+
+    def _custom_initialize_method(self):
```

Comment:
This seems to be a round-about way of doing this, and there is not an obvious API for how the user would setthe custom initialize method.

My first instinct is to just put a reference to the custom initialization method in the config block, and call it directly from there. I.e.

```
def initialize(self, *args, **kwargs):
    if self.config.custom_initializer is None:
        self._default_initialize(*args, **kwargs)
    else:
        self.config.custom_initializer(*args, **kwargs)
```

There is also a question of defining the standard arguments a custom initialization methods should expect.

Reply:
I think I would go one farther and just call the config entry `initializer`, and have it hold a callable.  The default value would be `SkeletonUnitModelData._default_initializer`.  I think this would work easier if the function signature was `(unit, *args, **kwargs)`, then `_default_initializer` could be a `@staticmethod` (and declared above the `CONFIG` declaration).  That would help with things like cloning / pickling, as the callable is no longer a bound method.

Reply:
Thanks for the suggestions @andrewlee94  and @jsiirola. I have updated as recommended. Thanks for @jsiirola for help with the staticmethod. I think it looks nice and clean now.